lib/writable.js

```diff
@@ -1,69 +1,252 @@
-
-Writable.prototype.destroy = function(error) {
-  throw error;
-};
-
-Writable.prototype.cork = function() {
-  this.corked++;
-};
-
-Writable.prototype.uncork = function() {
-  if (this.corked) this.corked--;
-};
-
-
-Writable.prototype.setDefaultEncoding = function(encoding) {
-  this.encoding = encoding;
-};
+const EventEmitter = require('events');
+
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_STREAM_DESTROYED,
+  ERR_UNKNOWN_ENCODING,
+  ERR_STREAM_NULL_VALUES,
+  ERR_STREAM_WRITE_AFTER_END,
+  ERR_METHODS_NOT_IMPLEMENTED
+} = require('./errors');
+
+const DEFAULT_BUFFER_SIZE = 16 * 1024; // 16Kb
+
+const errorSymbol = Symbol('error');
+const writeSymbol = Symbol('write');
+const validateChunkSymbol = Symbol('validateChunk');
+
+class WritableBuffer {
+  constructor(bufferSize) {
+    this.offset = 0;
+    this.queue = [];
+    this.writableLength = 0;
+    this.bufferSize = bufferSize;
+    this.buffer = Buffer.allocUnsafe(bufferSize);
+  }
+
+  writeChunk(chunk) {
+    chunk.copy(this.buffer, this.offset);
+    this.writableLength += chunk.length;
+    this.offset += chunk.length;
+  }
+
+  write(chunk, encoding) {
+    const remainingSize = this.bufferSize - this.offset;
+
+    if (chunk.length <= remainingSize) {
+      this.writeChunk(chunk, encoding);
+    } else {
+      const currentBufferChunk = chunk.slice(0, this.remainingSize);
+      this.writeChunk(currentBufferChunk);
+      this.queue.push(this.buffer);
+      this.buffer = Buffer.allocUnsafe(this.bufferSize);
+      this.offset = 0;
+
+      const nextBufferChunk = chunk.slice(remainingSize);
+
+      if (nextBufferChunk.length < this.buffer.length) {
+        this.writeChunk(nextBufferChunk, encoding);
+      } else {
+        this.queue.push(nextBufferChunk);
+        this.writableLength += nextBufferChunk.length;
+      }
+    }
+  }
+
+  getBuffer() {
+    const bufferData = this.buffer.slice(0, this.offset);
+    this.queue.push(bufferData);
+    const ret = this.queue;
+    this.queue = [];
+    this.offset = 0;
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    final = null,
+    destroy = null,
+    defaultEncoding = 'utf-8',
+    bufferSize = DEFAULT_BUFFER_SIZE
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+    if (typeof final === 'function') this._final = final;
+    if (typeof destroy === 'function') this._destroy = destroy;
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.bufferSize = bufferSize;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(bufferSize);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked > 0) this.corked--;
+    if (this.corked === 0 && this.writableLength) {
```

Comment:
I'd prefer to be explicit
```suggestion
    if (this.corked === 0 && this.writableLength > 0) {
```